---
output: rmarkdown::github_document
bibliography: ./inst/REFERENCES.bib
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%",
    tidy = "styler"
)
library(fastGHS)
library(jointGHS)
library(MultiNetEnrich)
library(foreach)
```


<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
<!--[![codecov](https://codecov.io/gh/Camiling/JoStARS/branch/main/graph/badge.svg?token=QL5ZW3RQZD)](https://codecov.io/gh/Camiling/JoStARS) -->
<!--[![R build status](https://github.com/Camiling/JoStARS/workflows/R-CMD-check/badge.svg)](https://github.com/Camiling/JoStARS/actions) -->
<!-- badges: end -->

<!-- IF ON CRAN [![CRAN_Status_Badge](https://www.r-pkg.org/badges/version-last-release/shapr)]
[![CRAN_Downloads_Badge](https://cranlogs.r-pkg.org/badges/grand-total/shapr)]---->
<!--PAPER HERE [![DOI](https://joss.theoj.org/papers/10.21105/joss.02027/status.svg)]---->


# `MultiNetEnrich` <img src="man/figures/logo_enrich.png" align="right" height="150"/>


The `MultiNetEnrich` package implements functional network enrichment analysis for multiple data sets. The method takes a list of data matrices with a one-to-one correspondence between the variables, such as different clinical groups or tissue types with the same set of genes measured, or mRNA and protein encoded by the same genes. The method has two steps; (i) network estimation and (ii) enrichment analysis based on the results. In step (i), we use the joint graphical horseshoe for multiple network estimation (@lingjaerde2022scalable) to obtain one network per group. The method borrows strength across shared patterns while highlighting differences across networks, giving a reasonable basis for comparisons. In step (ii), we use ranked gene set enrichment analysis (GSEA) (@subramanian2005gene) on each network, where genes (or the proteins they encode) are ranked according to their node degree in the respective networks to identify gene sets that are over-represented by the most active or central genes or proteins in each group in terms of associations. To identify enriched gene sets, we walk through the ranked gene lists from top to bottom, investigating different gene sets and pathways from the Molecular Signatures Database (MSigDB).

![MultiNetEnrich workflow](man/figures/enrich_strategy.png)


## Installation

This package requires the network estimation packages `fastGHS`and `jointGHS`, as well as the `GSEA` package to be installed. To install the current development version of all packages, use

```{r, eval = FALSE}
remotes::install_github("camiling/fastGHS")
remotes::install_github("camiling/jointGHS")
remotes::install_github("GSEA-MSigDB/GSEA_R")
remotes::install_github("camiling/MultiNetEnrich")
```

If you would like to install all packages of the models we currently support, use

```{r, eval = FALSE}
remotes::install_github("camiling/MultiNetEnrich", dependencies = TRUE)
```


If you would also like to build and view the vignette locally, use 

```{r, eval = FALSE}
remotes::install_github("camiling/MultiNetEnrich", dependencies = TRUE, build_vignettes = TRUE)
browseVignettes("MultiNetEnrich")
```

You might need to run the script for GSEA manually once before the package can be used

```{r, eval = FALSE}
source(system.file('extdata', 'Run.GSEA.R', package = 'GSEA'))
```

## Specifying gene sets

The default gene sets are Wikipathways, PID, and Hallmark, but any gene set from MSigDb can be used. The gene sets to be used must be provided through the ``database`` argument as one or more strings in the form of the URLs listed [here](https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.5.1/).

## Example


The main function `MultiNetEnrich` takes a list of $K$ data matrices, each of dimension $n_k \times p$ where $n_k$ is the sample size of data set $k$, for which separate functional network enrichment is to be computed. Note that the column names must be gene names, so if you are using protein data please convert the column names to the id of the gene that encodes them. 

The following example applies the method to gene expression data from a study on acute leukemias; acute myeloid leukemia (AML) and acute lymphoblastic leukemia (ALL) (@armstrong2002mll). The output gives us the gene sets found to be significantly enriched at $p<0.05$ and false discovery rate $<25\%$. 

```{r, warning = FALSE,results=FALSE}
# Download data
dat.leuk = read.delim(file=
                        'https://data.broadinstitute.org/gsea-msigdb/gsea/dataset_files/Leukemia_collapsed_symbols.gct',
                      skip=2)
rownames(dat.leuk) = dat.leuk[,1]
dat.leuk = dat.leuk[,-c(1:2)]
# Separate ALL and AML
leuk.type = substr(colnames(dat.leuk),1,3)
dat.leuk.all = dat.leuk[,leuk.type=='ALL']
dat.leuk.aml = dat.leuk[,leuk.type=='AML']
# Select subset of transcriptomics for quick toy example
dat.leuk.all = t(dat.leuk.all[1:50,])
dat.leuk.aml = t(dat.leuk.aml[1:50,])
# Get gene names
gene.names = unlist(lapply(colnames(dat.leuk.all),
              FUN = function(s) strsplit(s,split='///')[[1]][1]))
colnames(dat.leuk.all) = gene.names 
colnames(dat.leuk.aml) = gene.names 
# Final data
dim(dat.leuk.all)
dim(dat.leuk.aml)
X = list(dat.leuk.all, dat.leuk.aml)

# Peform functional network enrichment
set.seed(1)
enrich.res = MultiNetEnrich(X)

```

We can inspect the enriched gene sets 

```{r}
# Look at enriched gene sets for AML
enrich.res$enrichment.list[[2]][,c('GS','SIZE','NES', 'NOM p-val','FDR q-val')]
```


The ranking of the genes can be printed for further inspection of the top hubs.

```{r,warning=FALSE}
# For AML
enrich.res$ranked.lists[[1]][1:5,] # Print top five genes
```


The estimated networks can be visualised with functions from the `network` and `ggnet2` libraries. 


```{r,fig.align='center', out.width='60%',results='hide',warning=FALSE}
set.seed(1234)
net.all <- network::network(enrich.res$networks[[1]])
net.aml <- network::network(enrich.res$networks[[2]])
g.all <- GGally::ggnet2(net.all, alpha = 0.5, color = "darkblue", label = colnames(X[[1]]))
g.aml <- GGally::ggnet2(net.aml, alpha = 0.5, color = "darkblue", label = colnames(X[[1]]))
ggpubr::ggarrange(g.all, g.aml, ncol = 2, nrow = 1)
```


## Contribution

All feedback and suggestions are very welcome. If you have any questions or comments, feel
free to open an issue [here](https://github.com/Camiling/MultiNetEnrich/issues). 


## References
